### DISCUSSION TITLE
Use of operatingState for device instance - where/how is it used?

### DISCUSSION THREAD

**Discussion by MarkVerheyen-eaton** [edgexfoundry/community#260] [Q&A]
The device instance has an operatingState field of "Up", "Down", and "Unknown" which would correlate with the present status of the typically southbound device. The question I have is (for example): If the device fails to respond to X requests then the device should be PATCHED in core-meta-data to indicate that the device is "Down". What service is responsible for updating the operatingState? Is it the device service or the application using the device service? In other words: Does the device service change the device instance state on communication failure or is it the responsibility of something else in the system to change the state? 

Are there any good examples of using the operatingState?

**Comment by cloudxxx8** [comment]
@MarkVerheyen-eaton yes, the Device Service has the responsibility to update OperatingState, so SDK provides the internal API to achieve that.
https://github.com/edgexfoundry/device-sdk-go/blob/decf7a4867f03271291a6465384c5ba130c9fec8/pkg/interfaces/service.go#L94

However, we don't have a good example from the current reference implementation in the open source community.
The IOTech commercial device services supports this feature like this:

1. If the auto event or command fail to reach the real device with the limited retry times, update the core-metadata operating state to `DOWN`
2. During the device operating state is DOWN, all the auto event and command of this device would stop working
3. After a configurable duration, Device Service set the device operating state to UP and re-trigger auto event and start handling commands
4. If the device is still unreachable, change the operating to DOWN again, or keep it as UP
